File: kotlin/src/test/kotlin/smile/AssociationTest.kt
/*
 * Copyright (c) 2010-2026 Haifeng Li. All rights reserved.
 *
 * SMILE is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SMILE is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SMILE. If not, see <https://www.gnu.org/licenses/>.
 */
package smile.association

import java.util.function.Supplier
import java.util.stream.Stream
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 *
 * @author Haifeng Li
 */
class AssociationTest {

    var itemsets = arrayOf<IntArray>(
        intArrayOf(1, 3),
        intArrayOf(2),
        intArrayOf(4),
        intArrayOf(2, 3, 4),
        intArrayOf(2, 3),
        intArrayOf(2, 3),
        intArrayOf(1, 2, 3, 4),
        intArrayOf(1, 3),
        intArrayOf(1, 2, 3),
        intArrayOf(1, 2, 3)
    )

    val streamSupplier: Supplier<Stream<IntArray>> = Supplier {
        Stream.of(*itemsets)
    }

    @Test
    fun testFPGrowth() {
        println("FP-Growth")
        val tree = fptree(3, streamSupplier)
        val results = fpgrowth(tree).toList()
        for (itemSet in results) {
            println(itemSet)
        }

        assertEquals(8, results.size)

        assertEquals(3, results.get(0).support)
        assertEquals(1, results.get(0).items.size)
        assertEquals(4, results.get(0).items[0])

        assertEquals(5, results.get(1).support)
        assertEquals(1, results.get(1).items.size)
        assertEquals(1, results.get(1).items[0])

        assertEquals(6, results.get(6).support)
        assertEquals(2, results.get(6).items.size)
        assertEquals(3, results.get(6).items[0])
        assertEquals(2, results.get(6).items[1])

        assertEquals(8, results.get(7).support)
        assertEquals(1, results.get(7).items.size)
        assertEquals(3, results.get(7).items[0])
    }

    @Test
    fun testARM() {
        println("ARM")
        val tree = fptree(3, streamSupplier)
        val rules = arm(0.5, tree).toList()
        for (rule in rules) {
            println(rule)
        }

        assertEquals(9, rules.size)

        assertEquals(0.6, rules.get(0).support(), 1E-2)
        assertEquals(0.75, rules.get(0).confidence(), 1E-2)
        assertEquals(1, rules.get(0).antecedent().size)
        assertEquals(3, rules.get(0).antecedent()[0])
        assertEquals(1, rules.get(0).consequent().size)
        assertEquals(2, rules.get(0).consequent()[0])

        assertEquals(0.3, rules.get(4).support(), 1E-2)
        assertEquals(0.6, rules.get(4).confidence(), 1E-2)
        assertEquals(1, rules.get(4).antecedent().size)
        assertEquals(1, rules.get(4).antecedent()[0])
        assertEquals(1, rules.get(4).consequent().size)
        assertEquals(2, rules.get(4).consequent()[0])

        assertEquals(0.3, rules.get(8).support(), 1E-2)
        assertEquals(0.6, rules.get(8).confidence(), 1E-2)
        assertEquals(1, rules.get(8).antecedent().size)
        assertEquals(1, rules.get(8).antecedent()[0])
        assertEquals(2, rules.get(8).consequent().size)
        assertEquals(3, rules.get(8).consequent()[0])
        assertEquals(2, rules.get(8).consequent()[1])
    }
}

File: clojure/test/smile/association_test.clj
;
;   SMILE is free software: you can redistribute it and/or modify it
;   under the terms of the GNU General Public License as published by
;   the Free Software Foundation, either version 3 of the License, or
;   (at your option) any later version.
;
;   SMILE is distributed in the hope that it will be useful, but
;   WITHOUT ANY WARRANTY; without even the implied warranty of
;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;   GNU General Public License for more details.
;
;   You should have received a copy of the GNU General Public License
;   along with SMILE. If not, see <https://www.gnu.org/licenses/>.

(ns smile.association-test
  (:require [clojure.test :refer :all]
            [smile.association :as a])
  (:import [smile.association FPTree]
           [java.util.stream Stream Collectors]))

(defn- make-int-array [items]
  (let [arr (java.lang.reflect.Array/newInstance Integer/TYPE (count items))]
    (doseq [[i v] (map-indexed vector items)]
      (java.lang.reflect.Array/setInt arr i v))
    arr))

(def itemsets
  (let [data [[1 3] [2] [4] [2 3 4] [2 3] [2 3] [1 2 3 4] [1 3] [1 2 3] [1 2 3]]
        arr (java.lang.reflect.Array/newInstance (type (make-int-array [1])) (count data))]
    (doseq [[i v] (map-indexed vector (map make-int-array data))]
      (java.lang.reflect.Array/set arr i v))
    arr))

(defn- stream-to-vec [stream]
  (.collect stream (Collectors/toList)))

(deftest fpgrowth
  (let [results (a/fpgrowth 3 itemsets)
        results-list (stream-to-vec results)]
    (println "FP-Growth")
    (doseq [itemset results-list]
      (println itemset))

    (is (= 8 (count results-list)))

    (is (= 3 (.support (nth results-list 0))))
    (is (= 1 (alength (.items (nth results-list 0)))))
    (is (= 4 (aget (.items (nth results-list 0)) 0)))

    (is (= 5 (.support (nth results-list 1))))
    (is (= 1 (alength (.items (nth results-list 1)))))
    (is (= 1 (aget (.items (nth results-list 1)) 0)))

    (is (= 6 (.support (nth results-list 6))))
    (is (= 2 (alength (.items (nth results-list 6)))))
    (is (= 3 (aget (.items (nth results-list 6)) 0)))
    (is (= 2 (aget (.items (nth results-list 6)) 1)))

    (is (= 8 (.support (nth results-list 7))))
    (is (= 1 (alength (.items (nth results-list 7)))))
    (is (= 3 (aget (.items (nth results-list 7)) 0)))))

(deftest arm
  (let [tree (FPTree/of 3 itemsets)
        rules (a/arm 0.5 tree)
        rules-list (stream-to-vec rules)]
    (println "ARM")
    (doseq [rule rules-list]
      (println rule))

    (is (= 9 (count rules-list)))

    (is (< 0.59 (.support (nth rules-list 0)) 0.61))
    (is (< 0.74 (.confidence (nth rules-list 0)) 0.76))
    (is (= 1 (alength (.antecedent (nth rules-list 0)))))
    (is (= 3 (aget (.antecedent (nth rules-list 0)) 0)))
    (is (= 1 (alength (.consequent (nth rules-list 0)))))
    (is (= 2 (aget (.consequent (nth rules-list 0)) 0)))

    (is (< 0.29 (.support (nth rules-list 4)) 0.31))
    (is (< 0.59 (.confidence (nth rules-list 4)) 0.61))
    (is (= 1 (alength (.antecedent (nth rules-list 4)))))
    (is (= 1 (aget (.antecedent (nth rules-list 4)) 0)))
    (is (= 1 (alength (.consequent (nth rules-list 4)))))
    (is (= 2 (aget (.consequent (nth rules-list 4)) 0)))

    (is (< 0.29 (.support (nth rules-list 8)) 0.31))
    (is (< 0.59 (.confidence (nth rules-list 8)) 0.61))
    (is (= 1 (alength (.antecedent (nth rules-list 8)))))
    (is (= 1 (aget (.antecedent (nth rules-list 8)) 0)))
    (is (= 2 (alength (.consequent (nth rules-list 8)))))
    (is (= 3 (aget (.consequent (nth rules-list 8)) 0)))
    (is (= 2 (aget (.consequent (nth rules-list 8)) 1)))))